src/main/java/org/aksw/gerbil/semantic/sameas/impl/cache/FileBasedCachingSameAsRetriever.java

```diff
@@ -111,62 +111,88 @@ protected FileBasedCachingSameAsRetriever(SameAsRetriever decoratedRetriever,
         // if the cache contains the uri, return the set or a set
         // containing only the uri (use the read mutex!!!)
         Set<String> result = null;
+        boolean uriIsCached = false;
         try {
             cacheReadMutex.acquire();
         } catch (InterruptedException e) {
             LOGGER.error("Exception while waiting for read mutex. Returning null.", e);
             return null;
         }
-        boolean uriIsCached = uriSetIdMapping.containsKey(uri);
-        if (uriIsCached) {
-            int setId = uriSetIdMapping.get(uri);
-            if (setId != ENTITY_NOT_FOUND) {
-                result = sets.get(setId);
+        try {
+            uriIsCached = uriSetIdMapping.containsKey(uri);
+            if (uriIsCached) {
+                int setId = uriSetIdMapping.get(uri);
+                if (setId != ENTITY_NOT_FOUND) {
+                    result = sets.get(setId);
+                }
             }
+        } finally {
+            cacheReadMutex.release();
         }
         // If the URI is not in the cache, or it has been cached but the result
         // is null and the request should be retried
         if (!uriIsCached || (uriIsCached && (result == null) && requestEntitiesNotFound)) {
-            cacheReadMutex.release();
-            result = decoratedRetriever.retrieveSameURIs(uri);
+            requestUri(uri);
```

Comment:
I think here the returned Uri set should be assigned result?

Reply:
Thanks. Fixed.